py/server/deephaven/experimental/keyed_record_adapter.py

```diff
@@ -0,0 +1,184 @@
+        # Index of current row (i.e. index into arrays of data retrieved from each DB column)
+        row_idx = 0
+        for row_key in record_data_row_keys:
+            # Find the data key corresponding to the DB row key from
+            # which this record was retrieved. Map the data key to the record.
+            data_key_idx: int = row_key_to_data_key_positional_index.get(row_key)
+            data_key_for_row = data_keys[data_key_idx]
+
+            # print(f'Building record for row_idx {row_idx}, row_key {row_key}, data_key {data_key_idx}/{data_key_for_row}')
+
+            row_data = list()
+            if self.is_single_key_col:
+                row_data.append(data_key_for_row)
+            else:
+                row_data.extend(data_key_for_row)
+
+            for col_arr in record_data_arrs:
+                # print(f'Adding data: {col_arr[row_idx]}')
+                row_data.append(col_arr[row_idx])
+
+            print(f'row_data for row {row_idx} is: {row_data}')
+            record = self.record_adapter_function(*row_data)
+
+            if self.is_single_key_col:
+                # Add the record to the results dictionary
+                result_records_dict[data_key_for_row] = record
+            else:
+                result_records_dict[data_key_idx] = record
+
+            row_idx += 1
+
+        if self.is_single_key_col:
+            return result_records_dict
+        else:
+            # Build a list with the results at the same indices as the input data keys
+            # (or None if no result was found for the requested key).
+            # This is because Python does not allow lists to be used as dictionary keys.
+            results_list: typing.List[T] = list()
+            for idx in range(0, len(data_keys)):
+                if idx in result_records_dict:
+                    results_list.append(result_records_dict[idx])
+                else:
+                    results_list.append(None)
+
+            return results_list
+
+    def _convert_keys_seq_for_java(self, data_keys: typing.Sequence, key_dtype: DType) -> typing.Sequence:
+        if type(data_keys) is numpy.ndarray:
+            # Some types (e.g. bool/DateTime) have special handling for numpy:
+            return dhnp._np_to_dh_seq(key_dtype, data_keys)
+        else:
+            return data_keys
+
+
+def make_record_adapter_with_constructor(table: Table,
+                                         record_adapter_type: typing.Type[T],
+                                         key_cols: [str, typing.List[str]],
+                                         data_cols: [str, typing.List[str]]) -> KeyedRecordAdapter[K, T]:
+    def constructor_adapter_function(*row_data):
+        # Create the record from the data key and row data
+        # The constructor MUST take the keys + the data as arguments, in order.
+        record = record_adapter_type.__new__(record_adapter_type)
+        record_adapter_type.__init__(record, *row_data)
+        return record
```

Comment:
Is there a reason that you aren't just doing
`return record_adapter_type(record, *row_data)`?  ... which could be simplified to a lambda.

![image](https://user-images.githubusercontent.com/5250374/206023554-07431a58-ad50-4376-af93-5804fac5a374.png)